activerecord/lib/active_record/autosave_association.rb

```diff
@@ -388,8 +388,10 @@ def save_collection_association(reflection)
               records -= records_to_destroy
             end
 
+            is_habtm = (reflection.parent_reflection && reflection.parent_reflection.macro == :has_and_belongs_to_many)
+
             records.each do |record|
-              next if record.destroyed?
+              next if record.destroyed? || (is_habtm && record.persisted? && !record.changed?)
```

Comment:
I'm really not a fan of this implementation. It feels quite non-local. Can we handle this closer to where the problem exists?